apps/yggdrasil_web/lib/yggdrasil_web/ensure_permission.ex

```diff
@@ -0,0 +1,77 @@
+defmodule YggdrasilWeb.EnsurePermission do
+  @moduledoc ~S"""
+  ## Overview
+  This plug takes in a connection and set of permission and ensures that all
+  permissions provided are present for the user on the connection.
+
+    import Yggdrasil.EnsurePermission
+
+    plug EnsurePermission, [foo: [:read]] when action in [:index]
+
+  ## Roles and Permissions 
+  The user's permissions are dictated by their role. Each role has a set
+  resources and each resource has a set of permissions. With this scheme
+  the res_perms passed in are in the form of:
+
+  [resource: [perm_set], resource: [perm_set]]
+
+  for as many resources and permissions that are needed. Then each resource is
+  processed in the list looking up the resource on the role, and then validating
+  that the set of perms passed in are contained in the set for the given resource
+  on the role.
+
+  If and only if all resources provided are found, and all permissions for each
+  resource are in present for that resource on the user's given role will this
+  plug process the request, otherwise it will send a 401 and stop processing
```

Comment:
This whole block reads awkwardly. Maybe:

```
If and only if all permissions for all resources are granted to the user will the plug allow the request to be processed. Otherwise a 401 status code will be returned immediately and processing will stop.
```